### PR TITLE
Use class_weight to weigh validation_data

### LINKED ISSUE #7985: Possible issue with metrics val_weighted_categorical_accuracy.
Hi @fchollet . First of all let me thank you for your amazing work on **Keras**. 

I'm using keras=2.0.8 on the tensorflow-gpu=1.3.0 backend.

I'm trying to implement the use of `class_weight` on `model.fit` for a multi-label classification task with metrics `categorical accuracy` and `weighted categorical accuracy`.

I noticed that the output of `val_categorical_accuracy` is always equal to `val_weighted_categorical_accuracy`, which I guess is an incorrect behaviour.

My code looks roughly like this:

```
#This model has five classes which have an unbalanced representation of samples. I took the most represented and I rapported the others to it, having value 1
class_weight {0 : 1, 1 : 1.566, 2 : 22.34, 3 : 12.356, 4 : 29.981}


X_full, Y_full = full_set.as_matrix(columns=listColumnsInputs), full_set.as_matrix(columns=listColumnsOutputs)
X_train, Y_train = train_set.as_matrix(columns=listColumnsInputs), train_set.as_matrix(columns=listColumnsOutputs)
X_test, Y_test = test_set.as_matrix(columns=listColumnsInputs), test_set.as_matrix(columns=listColumnsOutputs)

X_full = X_full.reshape(len(X_full), len(listColumnsInputs))
X_train = X_train.reshape(len(X_train), len(listColumnsInputs))
X_test = X_test.reshape(len(X_test), len(listColumnsInputs))

batch_size = 128
nb_classes = len(listColumnsOutputs)

model = Sequential()
num_h = 100
numerolayers = 3
loss_function =  'mean_squared_logarithmic_error'
optimizer_function = 'SGD'

num_epoch = 1000
regularizerl2 = .0
regularizerl1 = .0

model.add(Dense(num_h, input_dim=len(listColumnsInputs), kernel_initializer='uniform', activation='relu', kernel_regularizer=regularizers.l2(regularizerl2), activity_regularizer=regularizers.l1(regularizerl1)))
for x in range(1, numerolayers): model.add(Dense(num_h, kernel_initializer='uniform', activation='relu', kernel_regularizer=regularizers.l2(regularizerl2), activity_regularizer=regularizers.l1(regularizerl1)))

model.add(Dense(len(listColumnsOutputs), kernel_initializer='uniform', activation='softmax'))

model.summary()

model.compile(loss=loss_function,  optimizer=optimizer_function, weighted_metrics=['categorical_accuracy'],  metrics=['categorical_accuracy'])

history = model.fit(X_train, Y_train, class_weight=class_weight, batch_size=batch_size, epochs=num_epoch, verbose=1, validation_data=(X_test, Y_test), callbacks=[TensorBoard(log_dir='./logs', histogram_freq=25, write_graph=True, write_images=False)])

score = model.evaluate(X_test, Y_test, verbose=0)
```

Is this a small bug? Am I wrong? 

Thanks a lot for your time.

### PR DESCRIPTION
**Pull request by nicolewhite**
Fixes #7985

### CODE REVIEW

**Comment by nicolewhite** [comment]
The update to `test_random_normal` is unrelated but that test was flaky and was failing my builds.

**Comment by fchollet** [comment]
Thank you for the PR. But I'm not sure we should do this.

You'll notice that `evaluate` has a `sample_weight` argument but no `class_weight` argument. Validation evaluation in `fit` follows the same pattern. The reason is that class weighting is meant to help training (by modulating contributions of different classes), but not to affect evaluation. Evaluation, because it reflects production constraints, implies that labels are not available when making predictions, and thus can't be used to modulate evaluation metrics.

If you want to "cheat" anyway (you're working with validation data, so labels are available) you can use your labels to produce a `sample_weight` array and pass it to `evaluate` or as part of `validation_data` in `fit`.

**Comment by nicolewhite** [comment]
Yeah that's fair. I wrote this because of the confusion it caused in #7985, as you end up having the same value for weighted and unweighted validation metrics despite passing in `class_weight`. I did suggest passing in `sample_weight` as the third part of `validation_data`, which is a little inconvenient as you need to manage both class weights and sample weights at that point, but I agree with your points above nonetheless.

**Comment by ozabluda** [comment]
Maybe `validation_class_weight` should be introduced. It's absence is also confusing, for example:
 https://github.com/fchollet/keras/issues/5616#issuecomment-338431623

Ideally (I think) everything that takes `sample_weight` should also take `class_weight` for consistency.

I think by default `validation_class_weight` should be equal to `class_weight`, unless explicitly provided, maybe with a special value, which means 'unweighted'.

**Comment by nicolewhite** [comment]
Idk about a new variable. Perhaps it would be better to default to using `class_weight` on the validation data and allow people to not weigh their validation data by explicitly passing in `None` as the third part of `validation_data`, e.g. `validation_data=(x_test, y_test, None)`. Or something.

**Comment by ozabluda** [comment]
Maybe a tuple `(x_val, y_val, val_sample_weights, val_class_weight)` should be accepted, with `(x_val, y_val, val_sample_weights, None)` overriding?
